ark/jsonschema/array.ts

```diff
@@ -0,0 +1,123 @@
+import {
+	rootSchema,
+	type Intersection,
+	type Predicate,
+	type Traversal
+} from "@ark/schema"
+import { printable, throwParseError } from "@ark/util"
+import type { JsonSchema, Out, Type } from "arktype"
+import { parseJsonSchema } from "./json.ts"
+import { JsonSchemaScope } from "./scope.ts"
+
+const deepNormalize = (data: unknown): unknown =>
+	typeof data === "object" ?
+		data === null ? null
+		: Array.isArray(data) ? data.map(item => deepNormalize(item))
+		: Object.fromEntries(
+				Object.entries(data)
+					.map(([k, v]) => [k, deepNormalize(v)] as const)
+					.sort((l, r) => (l[0] > r[0] ? 1 : -1))
+			)
+	:	data
+
+const jsonSchemaArrayUniqueItemsValidator = (
+	array: readonly unknown[],
+	ctx: Traversal
+) => {
+	const seen: Record<string, true> = {}
+	const duplicates: unknown[] = []
+	for (const item of array) {
+		const stringified = JSON.stringify(deepNormalize(item))
+		if (stringified in seen) duplicates.push(item)
+		else seen[stringified] = true
+	}
+	return duplicates.length === 0 ?
+			true
+		:	ctx.reject({
+				expected: "unique array items",
+				actual: `duplicated at elements ${printable(duplicates)}`
+			})
+}
+
+const arrayContainsItemMatchingSchema = (schema: Type) => {
+	const jsonSchemaArrayContainsValidator = (
+		array: readonly unknown[],
+		ctx: Traversal
+	) =>
+		array.some(item => schema.allows(item)) === true ?
+			true
+		:	ctx.reject({
+				expected: `an array containing at least one item matching 'contains' schema of ${schema.description}`,
```

Comment:
Again very good errors overall. maybe we can pare this down a bit since we already use the word compare to just:


`... at least one item satisfying ${schema.expression}`

Reply:
I shortened slightly within dd563118